rr_evgp/src/costmap_plugins/track_closing_layer.cpp

```diff
@@ -0,0 +1,248 @@
+        cv::Point br = endpnt + half_size_point;
+        tl.x = std::max(tl.x, 0);
+        tl.y = std::max(tl.y, 0);
+        br.x = std::min(br.x, contour_img.cols - 1);
+        br.y = std::min(br.y, contour_img.rows - 1);
+
+        // cv::rectangle(debug_img, tl, br, cv::Scalar(255, 0, 0), 1);
+
+        // Find surrounding points
+        cv::Mat endpnt_box = contour_img(cv::Rect(tl, br));
+        std::vector<cv::Point> box_points;
+        cv::findNonZero(endpnt_box, box_points);
+
+        // Get regression line
+        cv::Vec4f line;
+        cv::fitLine(box_points, line, CV_DIST_L2, 0, 0.01, 0.01);
+        double vx = line(0);
+        double vy = line(1);
+        double x = line(2) + tl.x;
+        double y = line(3) + tl.y;
+
+        // Get direction of line from end point
+        cv::Point2d new_endpnt = intersect_point_line(endpnt, cv::Point2d(x, y), cv::Point2d(x + vx, y + vy), false);
+
+        cv::Vec2d avg_pnt;
+        for (const cv::Point& pnt : box_points) {
+            avg_pnt(0) += pnt.x + tl.x;
+            avg_pnt(1) += pnt.y + tl.y;
+        }
+        avg_pnt /= static_cast<double>(box_points.size());
+
+        cv::Vec2d v(vx, vy);
+        double projection = (cv::Vec2d(new_endpnt) - avg_pnt).dot(v);
+        if (projection < 0) {
+            vx *= -1;
+            vy *= -1;
+        }
+
+        // cv::circle(debug_img, new_endpnt, 2, cv::Scalar(0, 255, 0));
+        cv::Point2d point;
+        point.x = new_endpnt.x + extrapolate_distance_ * vx;
+        point.y = new_endpnt.y + extrapolate_distance_ * vy;
+        endpnt_line.emplace_back(new_endpnt, point, vx, vy);
+    }
+
+    // Connect end points
+    cv::Mat walls(mat_grid.rows, mat_grid.cols, CV_8UC1, cv::Scalar(0));
+    std::valarray<bool> have_connection(false, endpnt_line.size());
+    for (size_t i = 0; i < endpnt_line.size(); i++) {
+        const auto& curr_line_context = endpnt_line.at(i);
+        for (size_t j = i + 1; j < endpnt_line.size(); j++) {
+            const auto& comp_line_context = endpnt_line.at(j);
+            if (IsConnectable(curr_line_context, comp_line_context)) {
+                have_connection[i] = true;
+                have_connection[j] = true;
+                const auto& point_a = std::get<0>(curr_line_context);
+                const auto& point_b = std::get<0>(comp_line_context);
+                // cv::line(debug_img, point_a, point_b, cv::Scalar(255, 0, 255));
+                cv::line(walls, point_a, point_b, cv::Scalar(255));
+            }
+        }
+
+        if (!have_connection[i]) {
+            const auto& endpnt = std::get<0>(curr_line_context);
+            const auto& extrapolated_pnt = std::get<1>(curr_line_context);
+            // cv::line(debug_img, endpnt, extrapolated_pnt, cv::Scalar(0, 0, 255));
+            cv::line(walls, endpnt, extrapolated_pnt, cv::Scalar(255));
+        }
+    }
+
+    // Make map from image
+    cv::morphologyEx(walls, walls, cv::MORPH_DILATE, dilate_kernel);
```

Comment:
I think the dilation here should use a smaller kernel than what is used before skeletonization. Or at least we should be able to change their sizes independently